overlord/ifacestate/handlers.go

```diff
@@ -425,6 +425,20 @@ func (m *InterfaceManager) undoSetupProfiles(task *state.Task, tomb *tomb.Tomb)
 	}
 	snapName := snapsup.InstanceName()
 
+	// The previous task's undo (link-snap) may have triggered a restart, if this
+	// is the case we can only proceed once the restart has happened or we
+	// may be invoking tools (like apparmor) from the wrong snapd. We set
+	// the default to true as we cannot set it otherwise since the change will
+	// always have been created by the old snapd (that may not have "finish-restart")
+	logger.Debugf("finish restart from undoLinkSnap")
+	finishOpts := snapstate.FinishRestartOptions{
+		// Only default to true for snapd snap
+		FinishRestartDefault: snapsup.Type == snap.TypeSnapd,
+	}
+	if err := snapstateFinishRestart(task, snapsup, finishOpts); err != nil {
```

Comment:
@bboozzoo FinishRestart has some code related to snap-failure which I'm not sure is relevant or not in this case?

Reply:
it's triggered when starting snapd directly from snap-failure with SNAPD_REVERT_TO_REV set in environment.

Reply:
I was unsure if we need to differentiate the direction of the FinishRestart to avoid doing some action